tests/apollo/util/bft.py

```diff
@@ -586,13 +586,18 @@ def force_quorum_including_replica(self, replica_id, primary=0):
                     nursery.start_soon(self.wait_for_checkpoint, replica_id,
                             checkpoint_num)
 
-    async def wait_for_checkpoint(self, replica_id, checkpoint_num):
+    async def wait_for_checkpoint(self, replica_id, expected_checkpoint_num=None):
+        """
+        Wait for a single replica to reach the expected_checkpoint_num.
+        If none is provided, return the last stored checkpoint.
+        """
         key = ['bc_state_transfer', 'Gauges', 'last_stored_checkpoint']
         while True:
```

Comment:
One thing I noticed is that we never actually give up on waiting for the checkpoint to complete. Maybe we should have some upper bound (a few minutes for that, where the test fails from a `trio.fail_after` call). I suppose it's also fine to leave it like this if the test should timeout at a higher level. In fact, that's likely the right thing to do, allow the failure wrapper at the caller since the functions compose. We don't actually do that checking in `fill_and_wait_for_checkpoint` or `test_checkpoint_creation` though.

Reply:
Thanks Andrew, great catch! I've added a `trio.fail_after` within the `wait_for_checkpoint()` method as it can be used anywhere in the code. This also avoids the clutter of doing `trio.fail_after` every time we use it...